crates/autopilot/src/domain/settlement/solution/mod.rs

```diff
@@ -0,0 +1,280 @@
+        self.trades
+            .iter()
+            .map(|trade| trade.native_fee(prices))
+            .sum()
+    }
+
+    pub fn new(
+        calldata: &eth::Calldata,
+        domain_separator: &eth::DomainSeparator,
+    ) -> Result<Self, Error> {
+        let tokenized::Tokenized {
+            tokens,
+            clearing_prices,
+            trades: decoded_trades,
+            interactions: _interactions,
+            auction_id,
+        } = tokenized::Tokenized::new(calldata)?;
+
+        let mut trades = Vec::with_capacity(decoded_trades.len());
+        for trade in decoded_trades {
+            let flags = tokenized::TradeFlags(trade.8);
+            let sell_token_index = trade.0.as_usize();
+            let buy_token_index = trade.1.as_usize();
+            let sell_token = tokens[sell_token_index];
+            let buy_token = tokens[buy_token_index];
+            let uniform_sell_token_index = tokens
+                .iter()
+                .position(|token| token == &sell_token)
+                .unwrap();
+            let uniform_buy_token_index =
+                tokens.iter().position(|token| token == &buy_token).unwrap();
+            trades.push(trade::Trade::new(
+                tokenized::order_uid(&trade, &tokens, domain_separator)
+                    .map_err(|err| Error::OrderUidRecover(err, auction_id))?,
+                eth::Asset {
+                    token: sell_token.into(),
+                    amount: trade.3.into(),
+                },
+                eth::Asset {
+                    token: buy_token.into(),
+                    amount: trade.4.into(),
+                },
+                flags.side(),
+                trade.9.into(),
+                trade::Prices {
+                    uniform: trade::ClearingPrices {
+                        sell: clearing_prices[uniform_sell_token_index],
+                        buy: clearing_prices[uniform_buy_token_index],
+                    },
+                    custom: trade::ClearingPrices {
+                        sell: clearing_prices[sell_token_index],
+                        buy: clearing_prices[buy_token_index],
+                    },
+                },
+            ));
+        }
+
+        Ok(Self { trades, auction_id })
+    }
+}
+
+pub mod error {
+    use super::*;
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum Error {
+        #[error(transparent)]
+        Decoding(#[from] tokenized::error::Decoding),
+        #[error("failed to recover order uid {0} for auction {1}")]
+        OrderUidRecover(tokenized::error::Uid, auction::Id),
+    }
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum Score {
+        #[error(transparent)]
+        Zero(#[from] competition::ZeroScore),
+        #[error("missing native price for token {0:?}")]
+        MissingPrice(eth::TokenAddress),
+        #[error(transparent)]
+        Math(trade::error::Math),
```

Comment:
nit, docs for the error types explaining why e.g. Zero or MissingPrice is a problem.

Reply:
added